### DISCUSSION TITLE
RuntimeError: Chroma is running in http-only client mode

### DISCUSSION THREAD

**Discussion by Andrej0909** [imartinez/privateGPT#902] [Q&A]
I got this error message and I just don't get the solution how to fix it:

Traceback (most recent call last):
  File "c:\Users\Administrator\Desktop\Chatgpt\A320privatGPTv5\ingest.py", line 166, in <module>
    main()
  File "c:\Users\Administrator\Desktop\Chatgpt\A320privatGPTv5\ingest.py", line 158, in main
    db = Chroma.from_documents(texts, embeddings, persist_directory=persist_directory, client_settings=CHROMA_SETTINGS)
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Program Files\Python311\Lib\site-packages\langchain\vectorstores\chroma.py", line 578, in from_documents
    return cls.from_texts(
           ^^^^^^^^^^^^^^^
  File "C:\Program Files\Python311\Lib\site-packages\langchain\vectorstores\chroma.py", line 533, in from_texts
    chroma_collection = cls(
                        ^^^^
  File "C:\Program Files\Python311\Lib\site-packages\langchain\vectorstores\chroma.py", line 109, in __init__
    self._client = chromadb.Client(_client_settings)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Program Files\Python311\Lib\site-packages\chromadb\__init__.py", line 107, in Client
    system = System(settings)
             ^^^^^^^^^^^^^^^^
  File "C:\Program Files\Python311\Lib\site-packages\chromadb\config.py", line 168, in __init__
    raise RuntimeError(
RuntimeError: Chroma is running in http-only client mode, and can only be run with 'chromadb.api.fastapi.FastAPI' as the chroma_api_impl.             see https://docs.trychroma.com/usage-guide?lang=py#using-the-python-http-only-client for more information.

Does anybody know the solution?
Thx!

**Comment by Andrej0909** [comment]
The issue:

Starting chromadb 0.40 the chroma_db_impl is no longer a supported parameter, it uses sqlite instead.

Removing the line
chroma_db_impl="duckdb+parquet",
from langchain.vectorstores/chroma.py solves the issue, but the earlier DB cannot be used or migrated.

**Reply by whykay-01** [reply]
I have a similar issue, it looks like they depreciated code for that functionality, but forgot to update the docs. Let me know if there is a solution

**Comment by rithesh6** [comment]
> The issue:
> 
> Starting chromadb 0.40 the chroma_db_impl is no longer a supported parameter, it uses sqlite instead.
> 
> Removing the line chroma_db_impl="duckdb+parquet", from langchain.vectorstores/chroma.py solves the issue, but the earlier DB cannot be used or migrated.


I am exactly stuck at this error since 2 hours, removing the above mentioned line is not solving the error. Is there any other possible solution for it ? Thanks @Andrej0909

**Comment by logan-markewich** [comment]
I get this error with just 

```
import chromadb
chroma_client = chromadb.EphemeralClient()
```

**Comment by Byusa** [comment]
I am also stuck with this error: 
```curl
RuntimeError: Chroma is running in http-only client mode, and can only be run with 'chromadb.api.fastapi.FastAPI' as the chroma_api_impl.             see https://docs.trychroma.com/usage-guide?lang=py#using-the-python-http-only-client for more information.
```

I am using:
```py
import chromadb
chroma_client = chromadb.HttpClient(host='localhost', port=8000) 
```
And the code works well on my macbook but not on my windows or linux. Any help is appreciated.

**Reply by sarthak7399** [reply]
Hi,
I also tried various solutions for the same issue with ChromaDB from the internet while integrating the embeddings, but I couldn't succeed until I used an older version of ChromaDB.
Try using:
name = "chromadb"
version = "0.4.15"

**Comment by dbardi** [comment]
I am getting this error when running chroma:

`chroma run --path /mydata`

Error:
`RuntimeError: Chroma is running in http-only client mode, and can only be run with 'chromadb.api.fastapi.FastAPI' as the chroma_api_impl.`
`             see https://docs.trychroma.com/usage-guide?lang=py#using-the-python-http-only-client for more information.`

Error is thrown here: [main.py, lines 259 - 264]:

`            # The thin client is a system with only the API component`
`            if settings["chroma_api_impl"] != "chromadb.api.fastapi.FastAPI":`
`                raise RuntimeError(`
`                    "Chroma is running in http-only client mode, and can only be run with 'chromadb.api.fastapi.FastAPI' as the chroma_api_impl. \`
`            see https://docs.trychroma.com/usage-guide?lang=py#using-the-python-http-only-client for more information."`
`                )`

Update this [main.py line 89]
`chroma_api_impl: str = "chromadb.api.segment.SegmentAPI"`
to
`chroma_api_impl: str = "chromadb.api.fastapi.FastAPI"`

**Reply by Byusa** [reply]
You don't have both installed chromadb and chromadb-client. If so uninstall one and keep the other one.

**Reply by dbardi** [reply]
I fixed.. I removed chroma, chroma-hnswlib and chroma-client.
I reinstalled just chroma and chroma-hnswlib (no need for chroma-client) and problem solved.

The code change I recommended above is incorrect; you can pass the chroma_api_impl value in the settings.

`from chromadb import HttpClient`
`from chromadb.config import Settings`
`settings = Settings(chroma_api_impl="chromadb.api.fastapi.FastAPI")`
`client = HttpClient(host=host, port=port, settings=settings)`

**Comment by raghuveerawankar** [comment]
I faced the same error.
Let me explain my scenario.
I am using Jupyter Notebook using Anaconda Navigator. 
I used 
chroma_client = chromadb.HttpClient(host='localhost', port=8000) 

and then I run command
chroma run --path /mydata

and it wasn't accepting any other clients as the http-only mode got on.

Then I run following commands in terminal of Jupyter
1) python -m pip uninstall chromadb
2) python -m pip uninstall chromadb-client

Then I restart my PC and environment
Then I verified whether all chromadb files are removed or not by importing them in my code

Then I reinstalled chromadb and then the error got fixed

Steps: -
1) Uninstall chromadb and chromadb-client
2) Restart device
3) Reinstall chromadb

Special thanks to Mr.taz from Chromadb Discord Community for suggestion.
Join Chromadb Discord Community : - https://discord.com/invite/MMeYNTmh3x

**Reply by felix-dksh** [reply]
+1. Solved for the codes executed in notebook.

**Reply by Himanshu507** [reply]
+2 It really works man, Thank you.

**Reply by michaeleng7** [reply]
It's working, I can't believe it, thank you very much!

**Reply by RAW-si18** [reply]
Thank you bro... Yup this is working... I wasted my 3-4 hours

**Reply by furquanSal** [reply]
This approach was helpful for me as well. Thanks for sharing!

**Comment by Kamal02D** [comment]
in my case the issue was that i'm using a proxy , the issue was resolved after i disabled it

**Reply by raghuveerawankar** [reply]
That's great 
Good to know your problem is solved.

**Comment by manjory** [comment]
I faced the same error.. the issue with mine was I was doing it as pip install whereas it needed pip3... and then it kept saying: 
RuntimeError: Chroma is running in http-only client mode, and can only be run with 'chromadb.api.fastapi.FastAPI' as the chroma_api_impl. See https://docs.trychroma.com/usage-guide?lang=py#using-the-python-http-only-client for more information.
chroma_api_impl bugged me for a day, finally realized the correct command in the current version only supports --path so the final command that worked was:

chroma run --host localhost --port 8000 --path chromadb.api.fastapi.FastAPI --path .../db_path

compared to the suggestion I found at many places which was
chroma run --host localhost --port 8000 --chroma_api_impl chromadb.api.fastapi.FastAPI --path .../db_path (this was wrong)

**Reply by raghuveerawankar** [reply]
Thank you for finding out new way
please post it on DIscord community and share the same there as well
ChromaDB Discord : https://discord.com/invite/MMeYNTmh3x

**Comment by yahyaghani** [comment]
it seems to be a conflict between chromadb, chroma-client and chroma-hnswlib; uninstall all and install either the client or chromadb depending on your usecase. Can confirm This fixes the issue.

**Reply by davidtbo** [reply]
This is not the fix for me. I only have chromadb and chroma-hnswlib installed. The latter was installed by the former. My requirements only lists chromadb. Still searching...

**Comment by kanaad-lims** [comment]
I had a similar issue, it mostly happens if you have both `chromadb` as well as `chromadb-client` installed. The chromadb-client is lightweight and does not support the persistent client.

To fix it just uninstall chromadb-client either by `pip uninstall chromadb-client` if using pip, else `uv pip uninstall chromadb-client` if using uv package manager.

I also installed chromadb version 1.5.5 as i found it to be quite stable (current latest version is 1.5.9 at the time of writing this comment)

So this should solve the issue of *Chroma is running in http-only client mode* !
